android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt

```diff
@@ -0,0 +1,136 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.engine.ui.questionnaire
+
+import android.app.Activity
+import android.content.Intent
+import android.graphics.Bitmap
+import android.provider.MediaStore
+import android.view.View
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.lifecycleScope
+import com.bumptech.glide.Glide
+import com.google.android.fhir.datacapture.validation.ValidationResult
+import com.google.android.fhir.datacapture.views.QuestionnaireItemViewHolderDelegate
+import com.google.android.fhir.datacapture.views.QuestionnaireItemViewHolderFactory
+import com.google.android.fhir.datacapture.views.QuestionnaireItemViewItem
+import com.google.android.material.button.MaterialButton
+import id.zelory.compressor.Compressor.compress
+import id.zelory.compressor.constraint.size
+import id.zelory.compressor.extension
+import id.zelory.compressor.saveBitmap
+import java.io.File
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import org.hl7.fhir.r4.model.QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent
+import org.hl7.fhir.r4.model.StringType
+import org.smartregister.fhircore.engine.R
+import org.smartregister.fhircore.engine.util.extension.encodeToBase64
+import org.smartregister.fhircore.engine.util.extension.hide
+import org.smartregister.fhircore.engine.util.extension.show
+
+class CustomPhotoCaptureFactory(fragment: Fragment) :
+  QuestionnaireItemViewHolderFactory(R.layout.custom_photo_capture_layout) {
+
+  private lateinit var tvPrefix: TextView
+  private lateinit var tvHeader: TextView
+  private lateinit var ivThumbnail: ImageView
+  private lateinit var btnTakePhoto: MaterialButton
+  private lateinit var launchCamera: ActivityResultLauncher<Intent>
+  var questionnaireResponse = QuestionnaireResponseItemAnswerComponent()
+
+  init {
+    registerCameraIntent(fragment)
+  }
+
+  private fun registerCameraIntent(fragment: Fragment) {
+    with(fragment) {
+      launchCamera =
+        registerForActivityResult(StartActivityForResult()) { result ->
+          if (result.resultCode == Activity.RESULT_OK) {
+            val imageBitmap = result.data?.extras?.get(EXTRA_IMAGE) as Bitmap
+            ivThumbnail.apply {
+              Glide.with(fragment).load(imageBitmap).into(this)
+              show()
+            }
+            btnTakePhoto.text = getString(R.string.replace_photo)
+            lifecycleScope.launch(Dispatchers.Main) {
+              val imageFile =
+                async(Dispatchers.IO) {
```

Comment:
You can use `withContext` instead of async coroutine that is returning a result

Reply:
You can use the `Main` and `IO` Dispatchers provided by [DispatcherProvider](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/util/DispatcherProvider.kt) for ease of testing coroutine functions

Reply:
Okay